server/sonar-server/src/main/java/org/sonar/server/qualityprofile/QProfileRulesImpl.java

```diff
@@ -0,0 +1,139 @@
+    verifyNotBuiltIn(profile);
+
+    Set<RuleKey> ruleKeys = activations.stream().map(RuleActivation::getRuleKey).collect(MoreCollectors.toHashSet(activations.size()));
+    RuleActivationContext context = ruleActivator.createContextForUserProfile(dbSession, profile, ruleKeys);
+
+    List<ActiveRuleChange> changes = new ArrayList<>();
+    for (RuleActivation activation : activations) {
+      changes.addAll(ruleActivator.activate(dbSession, activation, context));
+    }
+    activeRuleIndexer.commitAndIndex(dbSession, changes);
+    return changes;
+  }
+
+  @Override
+  public BulkChangeResult bulkActivateAndCommit(DbSession dbSession, QProfileDto profile, RuleQuery ruleQuery, @Nullable String severity) {
+    verifyNotBuiltIn(profile);
+    return doBulk(dbSession, profile, ruleQuery, (context, ruleKey) -> {
+      RuleActivation activation = RuleActivation.create(ruleKey, severity, null);
+      return ruleActivator.activate(dbSession, activation, context);
+    });
+  }
+
+  @Override
+  public List<ActiveRuleChange> deactivateAndCommit(DbSession dbSession, QProfileDto profile, Collection<RuleKey> ruleKeys) {
+    verifyNotBuiltIn(profile);
+    RuleActivationContext context = ruleActivator.createContextForUserProfile(dbSession, profile, ruleKeys);
+
+    List<ActiveRuleChange> changes = new ArrayList<>();
+    for (RuleKey ruleKey : ruleKeys) {
+      changes.addAll(ruleActivator.deactivate(dbSession, context, ruleKey, false));
+    }
+    activeRuleIndexer.commitAndIndex(dbSession, changes);
+    return changes;
+  }
+
+  @Override
+  public BulkChangeResult bulkDeactivateAndCommit(DbSession dbSession, QProfileDto profile, RuleQuery ruleQuery) {
+    verifyNotBuiltIn(profile);
+    return doBulk(dbSession, profile, ruleQuery, (context, ruleKey) -> ruleActivator.deactivate(dbSession, context, ruleKey, false));
+  }
+
+  @Override
+  public List<ActiveRuleChange> deleteRule(DbSession dbSession, RuleDefinitionDto rule) {
+    List<ActiveRuleChange> changes = new ArrayList<>();
+    List<Integer> activeRuleIds = new ArrayList<>();
+    db.activeRuleDao().selectByRuleIdOfAllOrganizations(dbSession, rule.getId()).forEach(ar -> {
+      activeRuleIds.add(ar.getId());
+      changes.add(new ActiveRuleChange(ActiveRuleChange.Type.DEACTIVATED, ar));
+    });
+
+    db.activeRuleDao().deleteByIds(dbSession, activeRuleIds);
+    db.activeRuleDao().deleteParamsByActiveRuleIds(dbSession, activeRuleIds);
+
+    return changes;
+  }
+
+  private static void verifyNotBuiltIn(QProfileDto profile) {
+    checkArgument(!profile.isBuiltIn(), "The built-in profile %s is read-only and can't be updated", profile.getName());
+  }
+
+  private BulkChangeResult doBulk(DbSession dbSession, QProfileDto profile, RuleQuery ruleQuery, BiFunction<RuleActivationContext, RuleKey, List<ActiveRuleChange>> fn) {
+    BulkChangeResult result = new BulkChangeResult();
+    Collection<RuleKey> ruleKeys = Sets.newHashSet(ruleIndex.searchAll(ruleQuery));
```

Comment:
why make a copy of the result here?
I didn't find any place where this was mutated (would be bad anyway). Any other reason?

Reply:
It enforces the unicity of the `Iterator` returned by `ruleIndex.searchAll(ruleQuery)`. Do you think it should be dropped?